delegate/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+use anchor_lang::{prelude::*, InstructionData};
+use anchor_spl::token::{transfer, Transfer, Token, TokenAccount, Approve, approve};
+use solana_program::{instruction::Instruction, program_option::COption};
+
+// TODO update with correct ID
+declare_id!("2Ls5MquEmp42AXBxKXX3a9Gu54aPYYVC19tV7RCMKsTt");
+
+#[account]
+pub struct SwapInfo {
+    pub is_initialized: bool,
+    pub poster: Pubkey,
+    pub poster_sell_account: Pubkey,
```

Comment:
I think think you really need the buy account, maybe not even the sell account since these should be ATAs 

https://docs.rs/anchor-lang/0.25.0/anchor_lang/derive.Accounts.html

check the associated_token

Reply:
True I guess they could be associated accounts, good suggestion.

Reply:
Wait hang on, I don't that actually adds much actually. You'd still need to record what tokens are being bought and sold, so you're not saving any space as you'd have to write the mint addresses instead. Might as well be the actual accounts as you gain flexibility for no extra space no?

Reply:
Since you don't have amounts I'm assuming it's an NFT so it's will be amount of 1 and given the seller and the buyer you can use ATA to derive the accounts instead of using random ones. You get the benefit of not having to store them but also they can be created (init) with the same macro so the entire lifecycle can be handled by anchor. Unless I misunderstood your point?

Reply:
It does have amounts: the fields poster_sell_amount and poster_buy_amount.

Ah okay, so I would need to store the mint addresses for the buy and sell types, so I don't save any space in SwapInfo, but it does mean you don't have to create any empty accounts before doing the trade: understood!